### PR TITLE
C620

### PR DESCRIPTION
**Pull request by lindaxiang**

### CODE REVIEW

File: seq_tools/validation/c620_rg_id_in_bam_check.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+            header_array = header.decode('utf-8').rstrip().split('\n')
+
+            for line in header_array:
+                if not line.startswith("@RG"):
+                    continue
+                rg_array = line.rstrip().replace('\t', '\\t')
+
+                # get rg_id from BAM header
+                rg_id_in_bam = ':'.join([
+                    kv for kv in rg_array.split('\\t') if kv.startswith('ID:')
```
Comment (junjun-zhang) [inline seq_tools/validation/c620_rg_id_in_bam_check.py:59]: Just realized that there is no reason to replace `\t` to `\\t` and then use `\\t` to do split, simply just use `\t` to do split. I guess that was a copy-and-paste artifact.

File: seq_tools/validation/c620_rg_id_in_bam_check.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                if f not in rg_id_in_bams:
+                    rg_id_in_bams[f] = []
+
+                # don't need to check rg_id uniquessness as it's checked already elsewehere earlier
+                rg_id_in_bams[f].append(rg_id_in_bam)
+
+        read_groups = self.metadata['read_groups']
+
+        offending_ids = {}
+        rg_id_in_bam = []
```
Comment (junjun-zhang) [inline seq_tools/validation/c620_rg_id_in_bam_check.py:71]: no need

File: seq_tools/validation/c620_rg_id_in_bam_check.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+        for rg in read_groups:
+            filename = rg['file_r1']
+            if not filename.endswith('.bam'):
+                continue  # skip if not BAM
+
+            if rg.get('read_group_id_in_bam', None) is not None:
+                rg_id_in_bam.append(rg['read_group_id_in_bam'])
```
Comment (junjun-zhang) [inline seq_tools/validation/c620_rg_id_in_bam_check.py:78]: don't need to keep and track `read_group_id_in_bam`

File: seq_tools/validation/c620_rg_id_in_bam_check.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+                "in BAM file. Offending ID(s): %s" % '; '.join(msg)
+            self.message = message
+            self.logger.info(f'[{self.checker}] {message}')
+
+        elif rg_id_in_bam:
+            self.status = 'VALID'
+            message = "'read_group_id_in_bam' are provided in metadata and the current checker is skipped. Please refer to the next checker for further information."
+            self.message = message
+            self.logger.info(f'[{self.checker}] {message}')
+
```
Comment (junjun-zhang) [inline seq_tools/validation/c620_rg_id_in_bam_check.py:103]: don't need this block

Comment (lindaxiang) [reply seq_tools/validation/c620_rg_id_in_bam_check.py:103]: I feel it is not good to send the message `'submitter_read_group_id' in metadata matches RG ID in BAM check: VALID` as we do NOT check whether they match or not.

Comment (junjun-zhang) [reply seq_tools/validation/c620_rg_id_in_bam_check.py:103]: good point, but yes and no. Even when `elif rg_id_in_bam` is true, it does not mean `submitter_read_group_id` check has been skipped. It's possible when there are multiple read groups, some provided `read_group_id_in_bam` some did not.

A simple way to improve clarity here is to make the message more clear: `For a read group, when 'read_group_id_in_bam' is not provided, 'submitter_read_group_id' in metadata matches RG ID in BAM check: VALID`

This actually brings up a general question I started to realize when started to write validation checks for BAM. For a pure FASTQ submission, these checks are not applicable. So reporting the check result as `VALID` could be misleading. Maybe we should change the status `VALID` to `PASS` for all checks? Also, we can add BAM file name(s) in the message to be explicit about which BAMs have been validated.

Comment (lindaxiang) [reply seq_tools/validation/c620_rg_id_in_bam_check.py:103]: yeah, I agree. I think `PASS` is better than `VALID`

**Pull request merged**